src/ReachSets/reach.jl

```diff
@@ -44,7 +44,7 @@ Interface to reachability algorithms for an LTI system.
 A dictionary with available algorithms is available via
 `Reachability.available_algorithms`.
 """
-function reach(S::AbstractSystem,
+function reach(S::AbstractSystem, # Does AbstractSystem a superset of HybridSystem?
```

Comment:
Yes.

Reply:
should it be `Union{AbstractContinuousSystem, AbstractContinuousSystem}` ?

Reply:
I guess ```AbstractSystem``` = ```Union{AbstractContinuousSystem, AbstractDiscreteSystem}```

Reply:
Up to now we have always discretized the system in advance. Maybe for now we can stick to only `AbstractDiscreteSystem`.
Or  we pull the discretization inside, which might make sense for the hybrid setting where you probably want to discretize on demand (unreachable locations should not be discretized), in which case we can just use `AbstractContinuousSystem`.
At the moment we do not optimize, so anything is fine.

Reply:
indeed, we can keep just `AbstractDiscreteSystem`.

Reply:
I guess it is a good idea then, to have ```AbstractDiscreteSystem``` for the ```d_reach``` and ```AbstractContinuousSystem``` for the ```reach```

Reply:
No, `d_reach` (side note: what does the `d` stand for?) is the hybrid algorithm, so this should stay `HybridSystem`. `reach` takes a diskrete system, so this could become `AbstractDiscreteSystem`.

Reply:
> I believe reach takes only continuous systems.

Conceptually yes, but the algorithm works by discretizing and then working on the discrete system. The current implementation performs the discretization outside (in `solve.jl`), so what you get as input to `reach` is a discrete system already. This is what I tried to explain above.
We anyway will have to change this, because `solve` would now receive a `HybridSystem` (and thus the discretization would fail).

> d_reach means discrete reach

Because `reach` also takes a discretized system (see my comment above), I find this confusing. Maybe we should have longer names with clearer description.

Proposal:
* Rename `reach` to `reach_continuous`. It takes a continuous system and computes the continuous successors. For that it first discretizes the system (i.e., you have to move that code from `solve.jl` here).
* Add a function `reach_discrete`. It takes a flowpipe and a hybrid system and outputs the discrete successors.
* Rename `d_reach` to `reach_hybrid`. It takes a hybrid system computes the reachable states. For that it calls `reach_continuous` and `reach_discrete`.

CC @mforets

Reply:
JuliaReach supports only continuous systems now, why did you need discretization?

Reply:
Well, this is how the algorithm works :smiley:

Reply:
I guess you mean time discretization, while I mean discrete in terms of discrete systems with continuous states(or modes), where you`re doing discretization by time. But, anyway, It`s a good design point.

Reply:
Ah, now I see the confusion :grin:
Yes, the name `AbstractDiscreteSystem` may be confusing in the context of hybrid systems. It does not have anything to do with a graph structure, only that time is discretized.

Reply:
See [here](https://github.com/JuliaReach/MathematicalSystems.jl/blob/master/src/discrete.jl#L31).